speakcivi.php

```diff
@@ -131,7 +131,70 @@ function speakcivi_civicrm_tokenValues(&$values, $cids, $job = null, $tokens = a
   foreach ($cids as $cid) {
     $values[$cid]['speakcivi.confirmation_hash'] = sha1(CIVICRM_SITE_KEY . $cid);
   }
+
+  //  cargo cult wants me to use this too: || in_array('member_city', $tokens['speakcivi']
+  if (array_key_exists('city', $tokens['contact'])) {
+      speakcivi_member_city($values, $cids, findMailingLanguage($job));
+  }
+}
+
+function findMailingLanguage($job) {
+    $cache_key = "speakcivi:mailinglanguage:$job";
+    $language = Civi::cache()->get($cache_key);
+    if ($language) { 
+        return $language;
+    }
+    $dao = CRM_Core_DAO::executeQuery("
+SELECT m.language 
+FROM civicrm_mailing_job j 
+JOIN civicrm_mailing m ON m.id=j.mailing_id 
+WHERE j.id=$job
+"
+    );
+    $language = $dao->fetchValue();
+    Civi::cache()->set($cache_key, $language);
+    return $language;
+}
+
+
+function speakcivi_member_city(&$tokens, $cids, $language) {
+    $fallbacks = CRM_Speakcivi_Tools_Dictionary::fallbackCityValues();
+
+    foreach ( $cids as $contact_id ) {
+        $tokens[$contact_id]['city'] = $fallbacks[$language];
```

Comment:
Shouldn't this be set only if there is no current value? Or is it that the contact tokens are always set after the "hook tokens" and override them?